### PR TITLE
MPEG DASH Support (initial)

### PR DESCRIPTION
**Pull request by beardypig**
This PR adds support to MPEG DASH (mpd) streams (it **does not support DRM**). 

This is something I have been working on for a while (started 2 years ago) and I have just gotten round to getting it to a state that it should be OK to merge. 

`DASHStream.parse_manifest` is the method to use if you wanna using DASHStreams in your plugins, and if you have a stream URL that you wanna try out you can use the `dash://` prefix.

**Test URLs**

- http://www-itec.uni-klu.ac.at/ftp/datasets/DASHDataset2014/BigBuckBunny/15sec/BigBuckBunny_15s_simple_2014_05_09.mpd
- http://bitmovin-a.akamaihd.net/content/sintel/sintel.mpd
- http://bitmovin-a.akamaihd.net/content/MI201109210084_1/mpds/f08e80da-bf1d-4e3d-8899-f0f6155f6efa.mpd
- http://yt-dash-mse-test.commondatastorage.googleapis.com/media/car-20120827-manifest.mpd (many audio variations) 
- http://yt-dash-mse-test.commondatastorage.googleapis.com/media/oops_cenc-20121114-signedlicenseurl-manifest.mpd (DRMed, doesn't work)

**Known Issues**

For streams with separate audio and video tracks some versions of FFMPEG do not start playing the stream until it has completely downloaded - this is a problem for live streams and long VOD. I have not managed to work out the cause.

I would like some people to test it out and we can iron out the bugs as we go.

### CODE REVIEW

File: src/streamlink/stream/dash.py [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+
+    def reload(self):
+        if self.closed:
+            return
+
+        self.reader.buffer.wait_free()
+        self.logger.debug("Reloading manifest")
+        res = self.session.http.get(self.mpd.url, exception=StreamError)
+
+        self.mpd = MPD(http.xml(res, ignore_ns=True),
```
Comment (back-to) [inline src/streamlink/stream/dash.py]: > AttributeError: 'NoneType' object has no attribute 'xml'

should be `self.session`

File: src/streamlink/stream/dash.py [hunk 2]
```diff
@@ -0,0 +1,160 @@
+import itertools
+from streamlink import StreamError, PluginError
+from streamlink.compat import urlparse, urlunparse
+from streamlink.plugin.api import http
```
Comment (back-to) [inline src/streamlink/stream/dash.py]: **http** can be removed for `self.session`

File: src/streamlink/stream/dash.py (hunk 1 above)
Comment (beardypig) [reply src/streamlink/stream/dash.py]: This was working for me... :)

File: src/streamlink/stream/dash.py (hunk 2 above)
Comment (beardypig) [reply src/streamlink/stream/dash.py]: And this wasn't working... :)

File: src/streamlink/stream/dash.py (hunk 1 above)
Comment (beardypig) [reply src/streamlink/stream/dash.py]: But it is nicer like this, we should aim to remove all the globals :P

Comment (beardypig) [reply src/streamlink/stream/dash.py]: It should be an issue, don't know why it was working for me :)

**Comment by back-to** [comment]
## Livestreams

_session id's are removed from the gist files_

---

`https://zattoo.com`

- very laggy

https://gist.github.com/back-to/468452022daa888d60bd97c632905f94#file-zattoo
https://gist.github.com/back-to/468452022daa888d60bd97c632905f94#file-manifest-mpd-zattoo

---

**NSFW** `https://www.myfreecams.com/` **NSFW**

- a bit laggy
- `[stream.dash][debug] Reloading manifest` gets spammed on the server

https://gist.github.com/back-to/468452022daa888d60bd97c632905f94#file-mfc
https://gist.github.com/back-to/468452022daa888d60bd97c632905f94#file-manifest-mpd-mfc

---

Audio only / Radio won't work

```
streamlink dash://"http://a.files.bbci.co.uk/media/live/manifesto/audio/simulcast/dash/nonuk/dash_low/ak/bbc_radio_one.mpd" best -l debug
```

here are some changes so it works, of course it will need some more changes.  

https://gist.github.com/back-to/437c363f3daee740481e4f0ba4d8f320

but CTRL-C won't work correctly with this changes

```
^C
[stream.dash][debug] Closing worker thread
[stream.dash][debug] Closing writer thread
^C
Interrupted! Exiting...
[cli][info] Closing currently open stream...
```

**Comment by beardypig** [comment]
Thanks @back-to, good find for audio only :) I haven't found an audio only stream before. 

For myfreecams it looks like `minimumUpdatePeriod` is set to 0.5s which seems a bit crazy - perhaps I misinterpreted  the specs for MPD :D Probably need to add a back-off for that...

For the lagginess, not sure ... do you wanna add DASH support to the zattoo plugin on a fork, so I can test it out? Which version of ffmpeg were you using?

**Comment by back-to** [comment]
ffmpeg version 3.4.2

---

**zattoo**

diff
https://gist.github.com/back-to/d5770291f917f7df13247785cf90a1c1#file-zattoo-diff

and here is the new file
https://gist.githubusercontent.com/back-to/d5770291f917f7df13247785cf90a1c1/raw/d90971e2a85c8d205d5230b801389d15f4596e3e/zattoo.py

there are two `.mpd` urls they will be printed in the terminal

---

here are some radio streams

https://github.com/jnk22/kodinerds-iptv/blob/master/iptv/dash/dash.m3u

but they use all the same source

some of the other streams are offline

**Comment by beardypig** [comment]
Perfect, thanks :)

**Comment by beardypig** [comment]
I'm not sure what's up with the streams not stopping, that is something that needs some debugging :(

**Comment by beardypig** [comment]
fe10499 should make the streams stop really quick, need to test the full effect of it though. I didn't have chance to test zattoo yet, that'll be next week :)

**Comment by BZHDeveloper** [comment]
hum...
`dash://"http://das-q1.tf1.fr/2/USP-0x0/21/35/13542135/ssm/13542135.ism/13542135.mpd?vk=MTM1NDIxMzUubXBk" best -l debug`
doesn't work

> Exception in thread Thread-DASHStreamWorker:
Traceback (most recent call last):
  File "/usr/lib/python3.6/threading.py", line 916, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3/dist-packages/streamlink/stream/segmented.py", line 56, in run
    for segment in self.iter_segments():
  File "/usr/lib/python3/dist-packages/streamlink/stream/dash.py", line 60, in iter_segments
    for segment in representation.segments(init=init):
  File "/usr/lib/python3/dist-packages/streamlink/stream/dash_manifest.py", line 471, in segments
    **kwargs):
  File "/usr/lib/python3/dist-packages/streamlink/stream/dash_manifest.py", line 351, in segments
    for media_url, available_at in self.format_media(**kwargs):
  File "/usr/lib/python3/dist-packages/streamlink/stream/dash_manifest.py", line 416, in format_media
    yield self.make_url(self.media.format(**kwargs)), 0
KeyError: 'Number'

**Comment by beardypig** [comment]
I'll take a look at that one @BZHDeveloper, thanks for the example.

**Comment by beardypig** [comment]
@BZHDeveloper that stream seems to be working now :)

**Comment by melmorabity** [comment]
@beardypig I created a new branch on my fork, by merging your mpegdash branch and by updating my pluzz plugin to enable DASH stream support: https://github.com/melmorabity/streamlink/tree/mpegdash
* I had to retrieve the redirect URL of the MPD manifest file to ensure that fragment URLs are correctly built
* DASH streams start to play, then seem to hang (tested with both ffplay and mplayer, as well as with all available DASH streams); at this moment, I get « [stream.dash][debug] Reloading manifest » messages in debug mode, alternating with segment downloads:
https://paste.fedoraproject.org/paste/OKDjNHfbKzhjxyND5Ih-OA

**Comment by back-to** [comment]
For this url _(might be geo blocked sometimes)_

```
streamlink -l debug dash://"https://live1tv-a.akamaihd.net/dash-live11/streams/1tv/1tvdash.mpd" best --http-header "Referer=Referer: https://www.1tv.ru/iframed/eump-embedlive_remote.html" --http-header "User-Agent=Mozilla/5.0 (X11; Linux x86_64; rv:52.0) Gecko/20100101 Firefox/52.0" --http-header "Origin=https://www.1tv.ru"
```

here is the `.mpd` file https://gist.github.com/back-to/cbe4da6d2d424379be2e43bf47b7f5dd

**1. Error**
```
File "src/streamlink/stream/dash_manifest.py", line 312, in __init__
  self.segmentAlignment = self.attr(u"segmentAlignment", default=False, parser=MPDParsers.bool_str)
File "src/streamlink/stream/dash_manifest.py", line 127, in attr
  return parser(value)
File "src/streamlink/stream/dash_manifest.py", line 56, in bool_str
  raise MPDParsingError("bool must be true or false")
streamlink.stream.dash_manifest.MPDParsingError: bool must be true or false
```

**2. Error**

The `Number` must be filled with zeros to a length of 9

**incorrect**

```
[stream.dash][error] Failed to open segment ... _$Number%09d$.mp4:
Unable to open URL: ... _$Number%09d$.mp4
(400 Client Error: Bad Request for url: ... _$Number%09d$.mp4)
```

**correct** URL with Number = `... _000100834.mp4`

---

DRM video but no DRM error msg
> error: Unable to parse XML: unbound prefix: line 7, column 8 (b'<MPD mediaPresentationDuration="P ...) 

```
streamlink dash://http://yt-dash-mse-test.commondatastorage.googleapis.com/media/car_cenc-20120827-manifest_vidonly.mpd best -l debug
```

**Comment by beardypig** [comment]
@melmorabity how long did it play for?

@back-to I've fixed the first two issues but now I have another problem to fix :) The other playlist from Google actually seems to be invalid xml :P I'll add a work around though.

**Comment by melmorabity** [comment]
@beardypig it plays during 3 seconds at most

**Comment by beardypig** [comment]
@melmorabity, @back-to: I have added some support for `Segment@r` which is for repeating segments in a segment timeline - this should improve the situation for both `pluzz` and `zattoo`. However,  it is incomplete and there seems to be some discrepancy in the timings between repeating and non-repeating timelines - eg. for France2 the audio and video are out of sync ... I'm working on it, but I thought I'd share the progress :)

**Comment by beardypig** [comment]
Further to my last update, I have added an option to the `FFMPEGMuxer` to enable the `-copyts` option, which appear to fix the audio/video sync issue I had in `pluzz`. I still had issues with lagging/stuttering, but I'm not sure if that was just my connection or my proxy, or there is still an issue with the timing of the segments.

@melmorabity when you have time, could you test it out?

**Comment by melmorabity** [comment]
@beardypig this is definitevely better, DASH Pluzz streams seem to play without problem, thanks for your great job!!
I will test recording tomorrow with pluzz. I will also enable DASH support for some other plugins.

**Comment by beardypig** [comment]
@melmorabity awesome, thanks for the testing :)

**Comment by beardypig** [comment]
@back-to with these latest changes the `zattoo` mpd streams seem to work fine - I tested a couple for 10-15 mins and they played just fine, with no lagging or sync issues.

**Comment by Wishbringer** [comment]
I know CohhCarnage used it today, worked fine with the latest released Streamlink so Twitch must still support the old API, for the moment? Some viewers using the browser player had some white screen issues though. 😆

**Comment by gravyboat** [comment]
@Wishbringer Yeah I tested it on a stream that had DASH enabled and it seems like it just used HLS in my player.

**Comment by back-to** [comment]
dash stream of http://www.ceskatelevize.cz/ct24#live

mpd and error
https://gist.github.com/back-to/e4d2f32ca502954ea8119ff44cce78af

**Comment by beardypig** [comment]
Number and time! What is this madness?! I’ll check it out :)

**Comment by beardypig** [comment]
Added some tests and fixed some bugs. Started a bit of a re-organisation of the tests, which I will continue in #1646.

**Comment by back-to** [comment]
> zattoo

**works**

---

>  bbc radio 

**works**

---

> 1tv

**_works with errors_**

a lot of *404 Client Error*

streamlink is **to fast** with the download, the file does not exist at this time.

because of this, there are audio and video issues.

---

> http://www.ceskatelevize.cz/ct24#live

it works but it's laggy on some reloads

it's also laggy without streamlink just in vlc,
so it might be some bad configuration on there side.

---
the browser might ask for drm, but it's not required

https://onet100.vod.pl/i/news.100/__prognoza-pogody-na-majowke/44l2c4

mpd + error
https://gist.github.com/back-to/1c0e77c5580da4adf735b337caa31219

**Comment by beardypig** [comment]
The 1TV stream is set up to work based off the wallclock - there is a difference between the server clock and the user clock, they might be generating the fragments at the "wrong" time. It should be lagging by 40 seconds as suggested in their manifest, so it would have to be quite a drift... could also be a bug.

**Comment by beardypig** [comment]
@back-to pushed a fix for vod.pl, was missing `publishTime` - now it falls back to 1970-1-1 00:00:00.

**Comment by beardypig** [comment]
@back-to can you test 1TV again with the latest commit - I've switched over to `datetime` for the available at times, which should fix some clock discrepancies and should avoid any timezone issues. I was able to play the stream for a while without any 404s. If you still get problems, double check that your system clock is not fast and report back :) Thanks.

**Comment by Matesy** [comment]
@beardypig :- Hey Mate Can U help Me With Zee5 Premium Video?? This Seems To Be DRM Protected ..
Url :- https://www.zee5.com/movies/details/pad-man/0-0-movie_2077184162
MPD received In Network Tab... 
https://zee5vod.akamaized.net/drm/PRIORITY1080/HINDI_MOVIES/PADMAN_HINDI_MOVIE.mp4/audio/en/mp4a/init.mp4
Is Their Anyway To Get This Video??

**Comment by Matesy** [comment]
@beardypig :- Hey Mate Can U help Me With Zee5 Premium Video?? This Seems To Be DRM Protected ..
Url :- https://www.zee5.com/movies/details/pad-man/0-0-movie_2077184162
received In Network Tab...
https://zee5vod.akamaized.net/drm/PRIORITY1080/HINDI_MOVIES/PADMAN_HINDI_MOVIE.mp4/video/avc1/6/init.mp4
Is Their Anyway To Get This Video??

**Comment by gravyboat** [comment]
@Matesy We don't support anything that's DRM protected, it's too much work to circumvent (there are some command line options you can look at for passing cookies and the like). You're welcome to mess with it, but please don't comment on this thread regarding that as it's totally out of scope for what this PR is working on.

**Comment by beardypig** [comment]
@back-to I just pushed a change that should fix the issue you had with 1TV.ru.

**Comment by beardypig** [comment]
@gravyboat I feel like this branch is pretty close to being ready. We've fixed a bunch of bugs and it would be nice to get some of the plugins updated to use DASH too :)

**Comment by beardypig** [comment]
Pinging @mmetak as there is a new package dependency for the tests - `python-freezegun` (available for arch).

**Comment by mmetak** [comment]
@beardypig Added `python-isodate` to depends and `python-freezegun` to checkdepends.

**Comment by Twilight0** [comment]
Initial means that individual plugins also need update to output DASH streams, for instance youtube, is that correct? So facebook plugin can now be reinstated.

**Comment by beardypig** [comment]
@Twilight0, correct. It also means that it might not be complete - the spec is large and there are some cases I excluded for technical reasons and others that I missed.

**Comment by thinkpad4** [comment]
Is this working for Twitch yet?  If so, how to I enable it?

**Comment by beardypig** [comment]
@thinkpad4 unfortunately twitch is not using DASH, they have their own custom variant of HLS that they are using. One day we'll support it ;)

**Comment by thinkpad4** [comment]
I see, ok thanks

**Comment by gravyboat** [comment]
@thinkpad4 Just to expand on what @beardypig mentioned and in case anyone else stumbles in here, what they're basically doing is reducing the number of cached segments that are required before a stream will start playing in their web player (at least this was the case as of a few weeks ago when they first rolled this out). It's pretty disappointing as I was hoping this was DASH support as well but quickly learned otherwise.

**Comment by bastimeyer** [comment]
@gravyboat 
@thinkpad4 
https://medium.com/@periscopecode/introducing-lhls-media-streaming-eb6212948bef
It basically comes down to using "HTTP/1.1 Chunked Transfer Coding" on early/future segments, meaning that requests are being made ahead of time, waiting for the segments to become available and downloading them as soon as possible. The client doesn't have to wait for a segment to be finished and can consume the data immediately (pass it to the player/demuxer).
#1676

**Comment by KazukiMutou** [comment]
I would love for this functionality to come to Streamlink. I've been dropping Streamlink in favor of just watching on the twitch site with almost zero delay. However, I much rather prefer using Streamlink so it's a bit of a dilemma at the moment.

**Comment by bastimeyer** [comment]
Let's keep the low latency Twitch streaming out of this DASH thread here, as it's unrelated...
There's already a thread for this: #1676

**Comment by Ordval** [comment]
Are the Irish RTÉ streams DRM ?
I can open picture in VLC, audio in Media Player Classic.

**Comment by beardypig** [comment]
>Are the Irish RTÉ streams DRM ? ...

@Ordval, you’ll have to give a bit more detail than that :)

**Comment by Ordval** [comment]
I'm hesitant to give out a stream, but here it goes:
```
C:\Users\ORD>call streamlink.exe --http-header X-Forwarded-For=79.140.221.254 hls://"http://cdn.rasset.ie/hls-live/_definst_/rte2/rte2-720.m3u8" best
[console][info] Found matching plugin hls for URL hls://http://cdn.rasset.ie/hls-live/_definst_/rte2/rte2-720.m3u8
[console][info] Available streams: live (worst, best)
[console][info] Opening stream: live (hls)
[console][info] Starting player: "C:\Program Files (x86)\VideoLAN\VLC\vlc.exe"
C:\Users\Ordval> call streamlink.exe --http-header X-Forwarded-For=79.140.221.254 hls://"http://cdn.rasset.ie/hls-live/_definst_/rte2/rte2-720.m3u8" best
[console][info] Found matching plugin hls for URL hls://http://cdn.rasset.ie/hls-live/_definst_/rte2/rte2-720.m3u8
[console][info] Available streams: live (worst, best)
[console][info] Opening stream: live (hls)
[console][info] Starting player: "C:\Program Files (x86)\VideoLAN\VLC\vlc.exe"
```
no audio. Apparently also works without X-F .

**Comment by bastimeyer** [comment]
@Ordval You are commenting in the MPEG DASH pull request, which already got merged. Your issue is related to HLS streams, as you can read in the log output.

I'll lock this thread, since it's been merged and all of the discussion afterwards has been off topic... Please open a new issue, if you think one of the plugins isn't working correctly.

**Pull request merged**